### ISSUE TITLE
Can't handle cpath in user.lua?

### ISSUE THREAD

**Issue by robertlzj** [pkulchenko/ZeroBraneStudio#934]
I use [iuplua](http://iup.sourceforge.net/) written by [scuri](http://sourceforge.net/u/scuri/). The library  file's name has postfix of lua interpreter version like 'iuplua51', 'iuplua53'.
I want to expand "package.cpath" to add lua interpreter version support on required file name. Like "clibs53" from 'clibs' do, then in code could just write "require 'iuplua'".
I find "luabase.lua" does this, so I tried:
```diff
diff --git "a/C:\\Users\\Robert\\AppData\\Local\\Temp\\TortoiseGit\\luabase-af683dc.000.lua" "b/C:\\ProgramFiles\\ZeroBraneStudio\\interpreters\\luabase.lua"
index 3c27f792..106586e4 100644
--- "a/C:\\Users\\Robert\\AppData\\Local\\Temp\\TortoiseGit\\luabase-af683dc.000.lua"
+++ "b/C:\\ProgramFiles\\ZeroBraneStudio\\interpreters\\luabase.lua"
@@ -81,6 +81,11 @@ return {
       local cpath = os.getenv(envcpath)
       local clibs = string.format('/clibs%s/', version):gsub('%.','')
       if not cpath:find(clibs, 1, true) then cpath = cpath:gsub('/clibs/', clibs) end
+			-- add references to version-specific .dll files.
+			local repl=function(path,version)
+				return path..version..'/?.dll'..path..version..'/?'..version..'.dll'
+			end
+			cpath=string.gsub(cpath,"(;[^;]*/clibs)(%d+)/%?%.dll",repl)
       wx.wxSetEnv(envcpath, cpath)
     end
```
It works fine, but I think I should implement it on "user.lua", but I failed.
I can neither change "package.cpath" nor "wx.wxSetEnv" in "user.lua" to do this.

```lua
local envcpath = "LUA_CPATH"
local cpath
--cpath=package.cpath
cpath = os.getenv(envcpath)
local repl=function(path,version)
	return path..version..'/?.dll'..path..version..'/?'..version..'.dll'
end
cpath=string.gsub(cpath,"(;[^;]*/clibs)(%d+)/%?%.dll",repl)
wx.wxSetEnv(envcpath, cpath)
```

**Comment by pkulchenko** [comment]
I'm not sure why it's not working in your case, but setting `wx.wxSetEnv("LUA_CPATH", "foobar")` in `user.lua` works for me (I get `foobar` when I execute script with `print(package.cpath)` in it).

**Comment by pkulchenko** [comment]
> Seems the value of cpath in my code is wrong?

Could be; make sure you add to the value instead of setting it from scratch, as there may be references needed for the IDE itself.

> By the way, can I debug or use print in user.lua?

Yes, `ide:Print("something")` should work.

**Comment by robertlzj** [comment]
I print from `user.lua`.
```lua
ide:Print(os.getenv("LUA_CPATH"))  --;;C:\ProgramFiles\ZeroBraneStudio\bin/clibs/?.dll
ide:Print(wx.wxGetEnv("LUA_CPATH"))  --true	  ;;C:\ProgramFiles\ZeroBraneStudio\bin/clibs/?.dll
```
Which is different from the actual value in runtime `print(package.cpath)`.
Means the one in `user.lua` isn't the latest value, even not the result of `luabase.lua`.
Also, `string.gsub` in my code has nothing to change.

**Comment by pkulchenko** [comment]
You can use `onInterpreterLoad` callback to set `LUA_CPATH` values in a plugin or user.lua based on the interpreter value that is being loaded. Don't forget to change it back to the original value when a different interpreter is loaded.

**Comment by robertlzj** [comment]
Goal achieved, but the implement is ugly. Remarks bellow:
1. I tried to get the result of `envcpath`, if done, only `string.gsub` is needed.
But `envcpath` is handled in function `frun` from `luabase.lua`, and don't know how to get the final value of `LUA_CPATH` or `package.cpath`.
So, I copy some codes in `frun` to ’onInterpreterLoad‘  callback in my plugin to do the process to get my `envcpath`.
2. Don't know how to get `onInterpreterLoad` callback invoked from `user.lua`.
`interpreter.luaversion` is needed from callback.

Code in plugin:
```lua
P.onInterpreterLoad = function(self, interpreter)
	if interpreter.luaversion~='5.1' then
		local envcpath = "LUA_CPATH"
		local cpath = os.getenv(envcpath)
		local clibs = string.format('/clibs%s/', interpreter.luaversion):gsub('%.','')
		if not cpath:find(clibs, 1, true) then cpath = cpath:gsub('/clibs/', clibs) end
		-- add references to version-specific .dll files.
		local repl=function(path,version)
			return path..version..'/?.dll'..path..version..'/?'..version..'.dll'
		end
		cpath=string.gsub(cpath,"(;[^;]*/clibs)(%d+)/%?%.dll",repl)
		print(interpreter.luaversion,cpath)
		wx.wxSetEnv(envcpath, cpath)
	end
end
```

**Comment by pkulchenko** [comment]
@robertlzj, from `user.lua` you can use the following:

```
package {
  onInterpreterLoad = ...skipped rest...
}
```

See https://studio.zerobrane.com/doc-configuration#configuration-file-commands for details.

**Comment by robertlzj** [comment]
I found debug couldn't work - won't break, and auto end without 'debug session start' in 3~5 seconds.

> Make sure you add to the value instead of setting it from scratch, as there may be references needed for the IDE itself.
> Don't forget to change it back to the original value when a different interpreter is loaded.

I tried to use `onInterpreterClose` to restore `wx.wxSetEnv(envcpath, cpath_original)`. Doesn't work.
I just change from:

> ;;C:\ProgramFiles\ZeroBraneStudio\bin/clibs/?.dll

to:

> ;;C:\ProgramFiles\ZeroBraneStudio\bin/clibs/?.dll;C:\ProgramFiles\ZeroBraneStudio\bin/clibs53/?.dll;C:\ProgramFiles\ZeroBraneStudio\bin/clibs53/?53.dll

**Comment by pkulchenko** [comment]
Are you starting debugging with `F5/Debug` or do you run the script outside of the IDE? If the debugging doesn't work, print the value of `package.cpath` at the beginning of your script to see what paths are being referenced there.

**Comment by robertlzj** [comment]
I start debugging in IDE.
`package.cpath` is:
- Run with plugin. (Can't debug with plugin)
  > ;C:\ProgramFiles\ZeroBraneStudio\bin\?.dll;C:\ProgramFiles\ZeroBraneStudio\bin\..\lib\lua\5.3\?.dll;C:\ProgramFiles\ZeroBraneStudio\bin\loadall.dll;.\?.dll;C:\ProgramFiles\ZeroBraneStudio\bin/clibs/?.dll;C:\ProgramFiles\ZeroBraneStudio\bin/clibs53/?.dll;C:\ProgramFiles\ZeroBraneStudio\bin/clibs53/?53.dll

- Normal run without plugin.

  > ;C:\ProgramFiles\ZeroBraneStudio\bin\?.dll;C:\ProgramFiles\ZeroBraneStudio\bin\..\lib\lua\5.3\?.dll;C:\ProgramFiles\ZeroBraneStudio\bin\loadall.dll;.\?.dll;C:\ProgramFiles\ZeroBraneStudio\bin/clibs53/?.dll


- Normal debug without plugin.

  > C:\ProgramFiles\ZeroBraneStudio\bin/clibs53/?.dll;C:\ProgramFiles\ZeroBraneStudio\bin\?.dll;C:\ProgramFiles\ZeroBraneStudio\bin\..\lib\lua\5.3\?.dll;C:\ProgramFiles\ZeroBraneStudio\bin\loadall.dll;.\?.dll;;C:\ProgramFiles\ZeroBraneStudio\bin/clibs53/?.dll

I use Environment `LUA_CPATH_5_3` now, and works fine.
If it is inconvenient, abandon the troubleshooting is OK.

**Comment by robertlzj** [comment]
I start debugging in IDE.
`package.cpath` is:
- Run with plugin. (Can't debug with plugin)
  > ;C:\ProgramFiles\ZeroBraneStudio\bin\?.dll;C:\ProgramFiles\ZeroBraneStudio\bin\..\lib\lua\5.3\?.dll;C:\ProgramFiles\ZeroBraneStudio\bin\loadall.dll;.\?.dll;C:\ProgramFiles\ZeroBraneStudio\bin/clibs/?.dll;C:\ProgramFiles\ZeroBraneStudio\bin/clibs53/?.dll;C:\ProgramFiles\ZeroBraneStudio\bin/clibs53/?53.dll

- Normal run without plugin.

  > ;C:\ProgramFiles\ZeroBraneStudio\bin\?.dll;C:\ProgramFiles\ZeroBraneStudio\bin\..\lib\lua\5.3\?.dll;C:\ProgramFiles\ZeroBraneStudio\bin\loadall.dll;.\?.dll;C:\ProgramFiles\ZeroBraneStudio\bin/clibs53/?.dll


- Normal debug without plugin.

  > C:\ProgramFiles\ZeroBraneStudio\bin/clibs53/?.dll;C:\ProgramFiles\ZeroBraneStudio\bin\?.dll;C:\ProgramFiles\ZeroBraneStudio\bin\..\lib\lua\5.3\?.dll;C:\ProgramFiles\ZeroBraneStudio\bin\loadall.dll;.\?.dll;;C:\ProgramFiles\ZeroBraneStudio\bin/clibs53/?.dll

**Comment by robertlzj** [comment]
Now I use Environment `LUA_CPATH_5_3` with value "..\clibs53\?53.dll" (and "..\clibs53\?.dll" #935), without using plugin.
It works fine.
If it is inconvenient, abandon the troubleshooting is OK.

**Comment by robertlzj** [comment]
I re-install system and zeroBrane.
The code mentioned above I added to expand cpath from "clibs53\\?.dll" to "clibs53\\?53.dll" in "luabase.lua" directly, could work in both _run_ and _debug_ mode _now_.
But if I transplant it to _plugin_ (in "user.lua") it could only work in _run_ mode. And in **_debug_** mode, won't execute any user code, will get "Program completed in 3.12 seconds (pid: 13076)." without prompting "Debugging session started in '...my path..." in normal case.

In other words, `wx.wxSetEnv(envcpath, cpath)` didn't work in _debug_ mode in _plugin_?
